### PR TITLE
Update hardware requirements for running Oasis Node (consensus, ParaTimes, Damask, Cobalt)

### PR DESCRIPTION
**Pull request by gw0**
Changes:
- Reorganized "Hardware recommendations".
- Improve estimates of storage usage and growth.
- Add network consumption section.

### CODE REVIEW

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
The Oasis Network is composed of multiple classes of nodes and services such
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]: Capitalized Validator, Non-validator, Node terms here https://docs.oasis.dev/general/run-a-node/set-up-your-node/ are just because they are part of the title.

```suggestion
* consensus validator or non-validator node
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Emerald ParaTime compute or client node
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Emerald Web3 gateway with PostgreSQL
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Cipher ParaTime compute or client node
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
If you run out of memory or storage, the
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
Oasis node process will be forcefully killed. This could lead to the state
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
corruption of BadgerDB, losing stake and not participating in committees.
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
### CPU {#suggested-minimum-configurations}
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Consensus validator or non-validator node:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Emerald ParaTime compute node and all ParaTime client nodes:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Emerald Web3 gateway with PostgreSQL:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Cipher ParaTime compute node:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
Emerald ParaTime client node behind a public Emerald Web3 gateway).
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
The [Intel SGX] support is required if you want to run ParaTime compute nodes
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Consensus validator or non-validator node:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* All ParaTime compute or client nodes:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
During regular workload your node will operate with less than the minimum amount of
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
state sync, BadgerDB migration when upgrading a node to a new major version of the Oasis Core,
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Consensus validator or non-validator node:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Emerald ParaTime compute or client node:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Emerald Web3 gateway with PostgreSQL:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* Cipher ParaTime compute or client Node:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
The consensus layer and ParaTime layers accumulate state over time. The speed at
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
which the state grows depends on the number of network and ParaTime transactions.
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
For example, a consensus non-validator node on the Mainnet accumulated:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* 280 GB of consensus state in ~1 year between Apr 28, 2021 and Apr 11, 2022 (since the [Cobalt upgrade])
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
* 32 GB of consensus state in ~1 month since the [Damask upgrade]
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
For example, an Emerald client node on the Mainnet additionally accumulated:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
For example, an Emerald Web3 gateway with PostgreSQL encountered:
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]: I would rewrite this block into a tip like this:

```
:::tip

Dump & restore upgrades (e.g. [Damask upgrade], [Cobalt upgrade]) include state
wipes which will free the node storage. Historical state can be accessed by running
a separate archive node.

:::
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]: Try to avoid passive speech and gerunds.

```suggestion
You can configure your node _not to_ keep a complete state from the
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
genesis onward which will reduce the amount of storage required for the consensus
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
and ParaTime state.
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
To enable pruning of the ParaTime state set the
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
certain time points when huge bursts of transactions arrive, you need to assure that
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
To enable pruning of the consensus state set the
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]: I would use *node* in a personal context.

```suggestion
During regular workload your node will operate with the minimal CPU resources.
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
However, if put under heavy load it might require more cores/vCPUs (e.g. an
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
memory. However, at certain time points, it will absolutely require more
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
During regular workload your node will receive much less network traffic. However, at
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
it doesn't timeout.
```

**Review by matevz — CHANGES_REQUESTED** [review]
Stylistic changes. I didn't verify actual requirement numbers, but they seem reasonable :)

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (gw0) [reply docs/general/run-a-node/prerequisites/hardware-recommendations.md]: For consistency I used capitalized "Consensus" when it appear in the beginning of all bullet points.

Suggestion: It would have been enough to suggest these changes once and tell to apply it throughout the document.

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (gw0) [reply docs/general/run-a-node/prerequisites/hardware-recommendations.md]: To not go too much into details here, I decided to remove the mention of "BadgerDB".

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (gw0) [reply docs/general/run-a-node/prerequisites/hardware-recommendations.md]: Much better. :+1:

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (gw0) [reply docs/general/run-a-node/prerequisites/hardware-recommendations.md]: I described it in an even longer way: "depends on the number of transactions on the network and ParaTimes."

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (gw0) [reply docs/general/run-a-node/prerequisites/hardware-recommendations.md]: Much better. :+1:

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (gw0) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]: @matevz Which suggestion sounds better?

```suggestion
  * In addition to the storage requirements for consensus state, the Emerald ParaTime requires additional storage
  * Minimum: 400+ GB of SSD or NVMe fast storage
  * Recommended: 700+ GB of SSD or NVMe fast storage
```

```suggestion
  * In addition to the storage requirements for consensus state, it requires at least additional: 400+ GB of SSD or NVMe fast storage
  * In addition to the storage requirements for consensus state, we recommend additional: 700+ GB of SSD or NVMe fast storage
```

Comment (matevz) [reply docs/general/run-a-node/prerequisites/hardware-recommendations.md]: Maybe replace the initial bullet with

```suggestion
* Emerald ParaTime compute or client node (in addition to the consensus storage requirements):
  * Minimum: 400+ GB of SSD or NVMe fast storage
  * Recommended: 700+ GB of SSD or NVMe fast storage
```

And the same for Cipher below.

~Maybe move the Cipher next to Emerald so all ParaTimes are grouped together?~ nvm, I see that this order is consistent throughout the document, so this is fine.

Comment (gw0) [reply docs/general/run-a-node/prerequisites/hardware-recommendations.md]: Suggestions applied and merged with master. Waiting for approval...

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 2.0 GHz x86-64 CPU with 2 cores/vCPUs with
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 2.0 GHz x86-64 CPU with 2 cores/vCPUs
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 2.0 GHz x86-64 CPU with 2 cores/vCPUs with
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Minimum: 400 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 700 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Minimum: 400 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 700 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Minimum: 300 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 500 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Minimum: 200 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 300 GB of SSD or NVMe fast storage
```

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (matevz) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Minimum: 200 Mbps internet connection with low latency
```

**Review by matevz — APPROVED** [review]
I went through the figures and IMO we should remove all `+` for any value. It's obvious that any bound here is minimal and having more storage/memory/cpus/network will always be better.

File: docs/general/run-a-node/prerequisites/hardware-recommendations.md
Comment (gw0) [inline docs/general/run-a-node/prerequisites/hardware-recommendations.md]:
```suggestion
  * Recommended: 2.0 GHz x86-64 CPU with 4 cores/vCPUs with
```

**Pull request merged**
